jupyter-org-extensions.el

```diff
@@ -0,0 +1,394 @@
+  "Execute all the src blocks that start before point."
+  (interactive)
+  (let ((p (point)))
+    (save-excursion
+      (goto-char (point-min))
+      (while (and (org-babel-next-src-block) (< (point) p))
+        (org-babel-execute-src-block)))))
+
+;;;###autoload
+(defun jupyter-org-restart-and-execute-to-point ()
+  "Kill the kernel and run src-blocks to point."
+  (interactive)
+  ;; FIXME: restaring the kernel does not work
+  (call-interactively #'jupyter-repl-restart-kernel)
+  (jupyter-org-execute-to-point))
+
+;;;###autoload
+(defun jupyter-org-restart-kernel-execute-buffer ()
+  "Restart kernel and execute buffer."
+  (interactive)
+  ;; FIXME: restaring the kernel does not work
+  (call-interactively #'jupyter-repl-restart-kernel)
+  (org-babel-execute-buffer))
+
+;;;###autoload
+(defun jupyter-org-jump-to-block (&optional N)
+  "Jump to a block in the buffer.
+If narrowing is in effect, only a block in the narrowed region.
+Use a numeric prefix N to specify how many lines of context to use.
+Defaults to 3."
+  (interactive "p")
+  (let ((p '()))
+    (when (= 1 N) (setq N 3))
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward org-babel-src-block-regexp nil t)
+        (push (list (format "line %s:\n%s"
+                            (line-number-at-pos (match-beginning 0))
+                            (save-excursion
+                              (goto-char (match-beginning 0))
+                              (let ((s (point)))
+                                (forward-line N)
+                                (buffer-substring s (point)))))
+                    (line-number-at-pos (match-beginning 0)))
+              p)))
+    (ivy-read "block: " (reverse p)
+              :action (lambda (candidate)
+                        (goto-char (point-min))
+                        (forward-line (1- (second candidate)))
+                        (outline-show-entry)
+                        (recenter)))))
+
+;;;###autoload
+(defun jupyter-org-jump-to-visible-block ()
+  "Jump to a visible src block with avy."
+  (interactive)
+  (avy-with jupyter-org-jump-to-block
+            (avy-jump "#\\+begin_src" :beg (point-min) :end (point-max))))
+
+;;;###autoload
+(defun jupyter-org-edit-header ()b
+  "Edit the src-block header in the minibuffer."
+  (interactive)
+  (let ((src-info (org-babel-get-src-block-info 'light)))
+    (unless src-info
+      (error "Not in a source block"))
+    (let* ((header-start (sixth src-info))
+           (header-end (save-excursion (goto-char header-start)
+                                       (line-end-position))))
+      (setf (buffer-substring header-start header-end)
+            (read-string "Header: "
+                         (buffer-substring header-start header-end))))))
+
+(defun jupyter-org-select-block-and-results ()
+  "Return the region that covers a source block and its results (if any).
+
+Return the region as (REGION_START . REGION_END)"
```

Comment:
`BEGIN` seems better here since `org-element` already uses `:begin` for the beginning of a source block. Also we can just say `(BEGIN . END)` here since the region is already mentioned.

Also I think `contains` sounds more correct here than `covers` since the text that is the source block is contained in the buffer region between `(BEGIN . END)`. WDYT?